kernel: kmod-ext4: enable ACL and Security labels by default

leobsky opened the pull request:
Fixes docker error: `failed to register layer: lsetxattr security.capability /usr/bin/ping: operation not supported` 
Forum discussion:
https://forum.openwrt.org/t/docker-pull-fails-failed-to-register-layer-operation-not-supported/138253

vincejv commented:
Shouldn't this change be isolated to platforms that support docker? Like arm64 and x86/64 in particular?

leobsky commented:
> Shouldn't this change be isolated to platforms that support docker? Like arm64 and x86/64 in particular? (i might've missed other platforms, but those are the two big ones)

This change only take effect on on kmod-ext4 and not all target use ext4 for the firmware images.

hauke commented on package/kernel/linux/modules/fs.mk:
```diff
@@ -232,6 +232,8 @@ define KernelPackage/fs-ext4
 	CONFIG_EXT4_FS \
 	CONFIG_EXT4_ENCRYPTION=n \
+	CONFIG_EXT4_FS_POSIX_ACL=y \
+	CONFIG_EXT4_FS_SECURITY=y \
```
Some targets are building `CONFIG_EXT4_FS` into the kernel. There this package is not build. 
I think it would be better to move these two options to `target/linux/generic/config-*`. 

How much bigger is the kernel getting with these two options and their dependencies?

leobsky replied:
> How much bigger is the kernel getting with these two options and their dependencies?

differences are minimal
normal
```
ls -l --block-size=K
12950K Aug 25 16:44  openwrt-x86-64-generic-ext4-combined-efi.img.gz
12754K Aug 25 16:44  openwrt-x86-64-generic-ext4-combined.img.gz
5663K Aug 25 16:44  openwrt-x86-64-generic-ext4-rootfs.img.gz
6701K Aug 25 16:43  openwrt-x86-64-generic-kernel.bin
```
with these two options
```
ls -l --block-size=K
12947K Aug 25 17:18 openwrt-x86-64-generic-ext4-combined-efi.img.gz
12759K Aug 25 17:18 openwrt-x86-64-generic-ext4-combined.img.gz
5663K Aug 25 17:18 openwrt-x86-64-generic-ext4-rootfs.img.gz
6705K Aug 25 17:18 openwrt-x86-64-generic-kernel.bin
```



> Some targets are building `CONFIG_EXT4_FS` into the kernel. There this package is not build.
> I think it would be better to move these two options to `target/linux/generic/config-*`.

I push changes following your suggestion, also deactivated  those options from targets where are enabled.

hauke commented:
I was thinking about this again. In our buildbot we are building all kernel modules. this would also activate `CONFIG_FS_POSIX_ACL` for all targets including devices with 8MB flash and MIPS CPUs where you probably do not want to run docker on. 

I think this would be a better solution:
```
--- a/config/Config-kernel.in
+++ b/config/Config-kernel.in
@@ -1247,6 +1247,7 @@ config KERNEL_BTRFS_FS
 menu "Filesystem ACL and attr support options"
 	config USE_FS_ACL_ATTR
 		bool "Use filesystem ACL and attr support by default"
+		default y if !SMALL_FLASH
 		help
 		  Make using ACLs (e.g. POSIX ACL, NFSv4 ACL) the default
 		  for kernel and packages, except tmpfs, flash filesystems,
@@ -1415,15 +1416,19 @@ config KERNEL_LSM
 
 config KERNEL_EXT4_FS_SECURITY
 	bool "Ext4 Security Labels"
+	default y if !SMALL_FLASH
 
 config KERNEL_F2FS_FS_SECURITY
 	bool "F2FS Security Labels"
+	default y if !SMALL_FLASH
 
 config KERNEL_UBIFS_FS_SECURITY
 	bool "UBIFS Security Labels"
+	default y if !SMALL_FLASH
 
 config KERNEL_JFFS2_FS_SECURITY
 	bool "JFFS2 Security Labels"
+	default y if !SMALL_FLASH
 
 config KERNEL_WERROR
 	bool "Compile the kernel with warnings as errors"
```
This should activate the `*_FS_SECURITY` and `*_FS_ACL_ATTR` options for all targets with big flash memory like x86 and armsr.

Please remove all target specific options which activate any of the  `*_FS_SECURITY` and `*_FS_ACL_ATTR`  options for a target in the `config-6.6` file.

leobsky commented:
> This should activate the `*_FS_SECURITY` and `*_FS_ACL_ATTR` options for all targets with big flash memory like x86 and armsr.
> 
> Please remove all target specific options which activate any of the `*_FS_SECURITY` and `*_FS_ACL_ATTR` options for a target in the `config-6.6` file.

I tested generating a config with the one that are being used by the OpenWrt buildbot on ath79  generic and tiny

```
CONFIG_DEFAULT_TARGET_ath79_tiny=y

# Filesystem ACL and attr support options
#
# CONFIG_USE_FS_ACL_ATTR is not set
# CONFIG_KERNEL_FS_POSIX_ACL is not set
# CONFIG_KERNEL_BTRFS_FS_POSIX_ACL is not set
# CONFIG_KERNEL_EXT4_FS_POSIX_ACL is not set
# CONFIG_KERNEL_F2FS_FS_POSIX_ACL is not set
# CONFIG_KERNEL_JFFS2_FS_POSIX_ACL is not set
# CONFIG_KERNEL_TMPFS_POSIX_ACL is not set
# CONFIG_KERNEL_CIFS_ACL is not set
# CONFIG_KERNEL_HFS_FS_POSIX_ACL is not set
# CONFIG_KERNEL_HFSPLUS_FS_POSIX_ACL is not set
# CONFIG_KERNEL_NFS_ACL_SUPPORT is not set
# CONFIG_KERNEL_NFS_V3_ACL_SUPPORT is not set
# CONFIG_KERNEL_NFSD_V2_ACL_SUPPORT is not set
# CONFIG_KERNEL_NFSD_V3_ACL_SUPPORT is not set
# CONFIG_KERNEL_REISER_FS_POSIX_ACL is not set
# CONFIG_KERNEL_XFS_POSIX_ACL is not set
# CONFIG_KERNEL_JFS_POSIX_ACL is not set
# end of Filesystem ACL and attr support options

# CONFIG_KERNEL_DEVMEM is not set
# CONFIG_KERNEL_DEVKMEM is not set
CONFIG_KERNEL_SQUASHFS_FRAGMENT_CACHE_SIZE=3
# CONFIG_KERNEL_SQUASHFS_XATTR is not set
# CONFIG_KERNEL_CC_OPTIMIZE_FOR_PERFORMANCE is not set
CONFIG_KERNEL_CC_OPTIMIZE_FOR_SIZE=y
# CONFIG_KERNEL_AUDIT is not set
# CONFIG_KERNEL_SECURITY is not set
# CONFIG_KERNEL_SECURITY_NETWORK is not set
# CONFIG_KERNEL_SECURITY_SELINUX is not set
# CONFIG_KERNEL_EXT4_FS_SECURITY is not set
# CONFIG_KERNEL_F2FS_FS_SECURITY is not set
# CONFIG_KERNEL_UBIFS_FS_SECURITY is not set
# CONFIG_KERNEL_JFFS2_FS_SECURITY is not set
CONFIG_KERNEL_WERROR=y
# end of Kernel build options
```
```
CONFIG_DEFAULT_TARGET_ath79_generic=y

# Filesystem ACL and attr support options
#
CONFIG_USE_FS_ACL_ATTR=y
CONFIG_KERNEL_FS_POSIX_ACL=y
CONFIG_KERNEL_BTRFS_FS_POSIX_ACL=y
CONFIG_KERNEL_EXT4_FS_POSIX_ACL=y
# CONFIG_KERNEL_F2FS_FS_POSIX_ACL is not set
# CONFIG_KERNEL_JFFS2_FS_POSIX_ACL is not set
# CONFIG_KERNEL_TMPFS_POSIX_ACL is not set
CONFIG_KERNEL_CIFS_ACL=y
CONFIG_KERNEL_HFS_FS_POSIX_ACL=y
CONFIG_KERNEL_HFSPLUS_FS_POSIX_ACL=y
CONFIG_KERNEL_NFS_ACL_SUPPORT=y
# CONFIG_KERNEL_NFS_V3_ACL_SUPPORT is not set
# CONFIG_KERNEL_NFSD_V2_ACL_SUPPORT is not set
# CONFIG_KERNEL_NFSD_V3_ACL_SUPPORT is not set
CONFIG_KERNEL_REISER_FS_POSIX_ACL=y
CONFIG_KERNEL_XFS_POSIX_ACL=y
CONFIG_KERNEL_JFS_POSIX_ACL=y
# end of Filesystem ACL and attr support options

# CONFIG_KERNEL_DEVMEM is not set
# CONFIG_KERNEL_DEVKMEM is not set
CONFIG_KERNEL_SQUASHFS_FRAGMENT_CACHE_SIZE=3
# CONFIG_KERNEL_SQUASHFS_XATTR is not set
CONFIG_KERNEL_CC_OPTIMIZE_FOR_PERFORMANCE=y
# CONFIG_KERNEL_CC_OPTIMIZE_FOR_SIZE is not set
# CONFIG_KERNEL_AUDIT is not set
# CONFIG_KERNEL_SECURITY is not set
# CONFIG_KERNEL_SECURITY_NETWORK is not set
# CONFIG_KERNEL_SECURITY_SELINUX is not set
CONFIG_KERNEL_EXT4_FS_SECURITY=y
CONFIG_KERNEL_F2FS_FS_SECURITY=y
CONFIG_KERNEL_UBIFS_FS_SECURITY=y
CONFIG_KERNEL_JFFS2_FS_SECURITY=y
CONFIG_KERNEL_WERROR=y
# end of Kernel build options
```

hauke commented on target/linux/bcm27xx/bcm2708/config-6.6:
```diff
@@ -143,7 +143,6 @@ CONFIG_EDAC_SUPPORT=y
 CONFIG_EXCLUSIVE_SYSTEM_RAM=y
 CONFIG_EXT4_FS=y
 CONFIG_EXT4_FS_POSIX_ACL=y
```
Setting `CONFIG_EXT4_FS_POSIX_ACL` should not be needed any more. I think these are not small targets and `CONFIG_EXT4_FS_POSIX_ACL` should be set because `USE_FS_ACL_ATTR` is set now. 

Maybe your configuration is not updated correctly, please check with `./scripts/diffconfig.sh`  is these settings are set to the default value.

hauke replied:
You can remove all the `*FS_POSIX_ACL=y` options for the kernel configuration files and only keep the options deactivated in the generic kernel configuration.

leobsky replied:
> You can remove all the `*FS_POSIX_ACL=y` options for the kernel configuration files and only keep the options deactivated in the generic kernel configuration.

done all `*FS_POSIX_ACL=y` removed.



> Maybe your configuration is not updated correctly, please check with `./scripts/diffconfig.sh` is these settings are set to the default value.

On ath79 generic returned to the default, value not on tiny, but also happen without applying the patch all `procd-ujail` are missed.
diff --side
```
CONFIG_ALL_KMODS=y                                              CONFIG_ALL_KMODS=y
CONFIG_ALL_NONSHARED=y                                          CONFIG_ALL_NONSHARED=y
CONFIG_DEVEL=y                                                  CONFIG_DEVEL=y
CONFIG_TARGET_PER_DEVICE_ROOTFS=y                               CONFIG_TARGET_PER_DEVICE_ROOTFS=y
CONFIG_AUTOREMOVE=y                                             CONFIG_AUTOREMOVE=y
CONFIG_BPF_TOOLCHAIN_BUILD_LLVM=y                               CONFIG_BPF_TOOLCHAIN_BUILD_LLVM=y
# CONFIG_BPF_TOOLCHAIN_NONE is not set                          # CONFIG_BPF_TOOLCHAIN_NONE is not set
CONFIG_BUILDBOT=y                                               CONFIG_BUILDBOT=y
CONFIG_COLLECT_KERNEL_DEBUG=y                                   CONFIG_COLLECT_KERNEL_DEBUG=y
CONFIG_DEFAULT_procd-ujail=y                                  <
CONFIG_HAS_BPF_TOOLCHAIN=y                                      CONFIG_HAS_BPF_TOOLCHAIN=y
CONFIG_IB=y                                                     CONFIG_IB=y
CONFIG_JSON_CYCLONEDX_SBOM=y                                    CONFIG_JSON_CYCLONEDX_SBOM=y
CONFIG_KERNEL_BUILD_DOMAIN="buildhost"                          CONFIG_KERNEL_BUILD_DOMAIN="buildhost"
CONFIG_KERNEL_BUILD_USER="builder"                              CONFIG_KERNEL_BUILD_USER="builder"
CONFIG_KERNEL_DEBUG_INFO=y                                      CONFIG_KERNEL_DEBUG_INFO=y
CONFIG_KERNEL_DEBUG_INFO_REDUCED=y                              CONFIG_KERNEL_DEBUG_INFO_REDUCED=y
CONFIG_KERNEL_DEBUG_KERNEL=y                                    CONFIG_KERNEL_DEBUG_KERNEL=y
CONFIG_KERNEL_KALLSYMS=y                                        CONFIG_KERNEL_KALLSYMS=y
CONFIG_MAKE_TOOLCHAIN=y                                         CONFIG_MAKE_TOOLCHAIN=y
CONFIG_MODULE_DEFAULT_procd-ujail=y                           <
CONFIG_PACKAGE_libbpf=m                                         CONFIG_PACKAGE_libbpf=m
CONFIG_PACKAGE_libelf=m                                         CONFIG_PACKAGE_libelf=m
CONFIG_PACKAGE_libpcap=m                                        CONFIG_PACKAGE_libpcap=m
CONFIG_PACKAGE_libxdp=m                                         CONFIG_PACKAGE_libxdp=m
CONFIG_PACKAGE_qosify=m                                         CONFIG_PACKAGE_qosify=m
CONFIG_PACKAGE_tc-tiny=m                                        CONFIG_PACKAGE_tc-tiny=m
CONFIG_PACKAGE_xdp-filter=m                                     CONFIG_PACKAGE_xdp-filter=m
CONFIG_PACKAGE_xdp-loader=m                                     CONFIG_PACKAGE_xdp-loader=m
CONFIG_PACKAGE_xdpdump=m                                        CONFIG_PACKAGE_xdpdump=m
CONFIG_PACKAGE_zlib=m                                           CONFIG_PACKAGE_zlib=m
# CONFIG_PKG_ASLR_PIE_NONE is not set                           # CONFIG_PKG_ASLR_PIE_NONE is not set
CONFIG_PKG_ASLR_PIE_REGULAR=y                                   CONFIG_PKG_ASLR_PIE_REGULAR=y
CONFIG_REPRODUCIBLE_DEBUG_INFO=y                                CONFIG_REPRODUCIBLE_DEBUG_INFO=y
CONFIG_SDK=y                                                    CONFIG_SDK=y
CONFIG_SDK_LLVM_BPF=y                                           CONFIG_SDK_LLVM_BPF=y
CONFIG_TARGET_ALL_PROFILES=y                                    CONFIG_TARGET_ALL_PROFILES=y
CONFIG_USE_LLVM_BUILD=y                                         CONFIG_USE_LLVM_BUILD=y
CONFIG_DEFAULT_procd-ujail=y                                  <
CONFIG_MODULE_DEFAULT_procd-ujail=y                           <
# CONFIG_PACKAGE_procd-ujail is not set                       <
```

leobsky committed:
kernel: activate *_FS_SECURITY and *_FS_ACL_ATTR options for all big flash targets

This patch activate *_FS_SECURITY and *_FS_ACL_ATTR options for all big flash memory targets.
Fixes docker error: "failed to register layer: lsetxattr security.capability /usr/bin/ping: operation not supported"
Forum discussion:
https://forum.openwrt.org/t/docker-pull-fails-failed-to-register-layer-operation-not-supported/138253

Signed-off-by: Leo Barsky <leobrsky@proton.me>
Link: https://github.com/openwrt/openwrt/pull/16181
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

leobsky commented:
Updating because commit decd5e0cff5efc666327714e5b83b11ceb52b93e bring new config symbols

hauke committed:
kernel: Activate POSIX ACL for f2fs, jffs2 and tmpfs

When CONFIG_USE_FS_ACL_ATTR is set we will also activate POSIX ACL
support for the f2fs, jffs2 and tmpfs file system. This option is
activated on all targets with big flash.

Link: https://github.com/openwrt/openwrt/pull/16181
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>
